src/test/scala/strawman/collection/test/TraverseTest.scala

```diff
@@ -128,12 +128,16 @@ class TraverseTest {
     val xs3: immutable.List[String] = xs2
 
     val xs4 = immutable.TreeMap((1, "1"), (2, "2"))
-    val xs5 = flatCollect(xs4) { case (2, v) => immutable.List((v, v)) }
+    val xs5 = flatCollect(xs4) { case (2, v) => immutable.List((v, v)) }(immutable.TreeMap)
```

Comment:
So, this is the issue. Here the `TreeMap` factory hasn’t been inferred because of an ambiguity with `SortedSet`.

Here is what happens, if I understand correctly. We are looking for an implicit `BuildFrom[TreeMap[Int, String], (String, String), C]`. The two candidates we find are:
- `SortedSet.canBuildSortedIterable`, of type `[A] => BuildFrom[Any, A, SortedSet[A]]`
- `TreeMap.canBuildSortedMap`, of type `[K, V] => BuildFrom[Any, (K, V), TreeMap[K, V]]`

Why are these candidates ambiguous? Because none is more specific than the other: `TreeMap[?, ?]` and `SortedSet[?]` are not subtypes of each other.

Why are they both available in the implicit scope? The one from the companion object `TreeMap` is available because we are looking for an implicit `BuildFrom[TreeMap[Int, String], ?, ?]`. For the other one, I guess this is because somewhere in the hierarchy of `SortedMap` we have a `SortedSet` that shows up (`SortedMap[K, V] extends SortedOps[K, SortedSet, SortedMap[K, V]]`). Consequently, the companion object of `SortedSet` contributes to the implicit scope.

How to fix the problem? Unfortunately, I don’t think we can make the types of one of the two implicits more specific than the other (e.g. we can not make `SortedMap[K, V]` _extend_ `SortedSet[K]` because they extend `Iterable[(K, V)]` and `Iterable[K]`, respectively). One solution could be to **change the priority of implicits**: when exploring companion objects, the further we go in the hierarchy (ie when we explore companions of _ancestors_ of `TreeMap`), the lower the priority of their implicit definitions should be. Does that make sense? With this change, I don’t think we could break existing code. However, code that doesn’t compile now (because of an ambiguity) may compile after this change.